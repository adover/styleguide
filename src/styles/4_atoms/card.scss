@import '../1_settings/_config';
@import '../2_tools/mixins/apply-spacing';
@import '../2_tools/mixins/display-flex';

$card-content-margin-bottom: $spacing-2;

/// A white background, square padding and a drop shadow.
/// Initially appeared on the product stamp but can be used in multiple locations.
/// @param $background-color [white]
/// @param {Bool} $border [false] - Add a 1px solid porcelain border
/// @param {Bool | default | medium | large} $padding [default] - Apply default, medium, large or no (false) spacing
/// @param {Bool} $shadow [map-get($shadows, small)] - Apply a box-shadow
/// @param {Bool | default | medium | large} $rounded-corners [false] - Apply a variety of border-radius sizes
/// @name card-layout
/// @group atoms
@mixin card-layout(
	$background-color: white,
	$border: false,
	$padding: default,
	$shadow: map-get($shadows, small),
	$rounded-corners: false
) {
	background-color: $background-color;
	border: if($border, map-deep-get($config, borders, divider), none);
	box-shadow: if($shadow, $shadow, none);
	overflow: hidden;
	position: relative;

	@if $rounded-corners == large {
		border-radius: map-get($radii, lg);
	}

	@else if $rounded-corners == medium {
		border-radius: map-get($radii, md);
	}

	@else if $rounded-corners == default {
		border-radius: map-get($radii, default);
	}

	@else {
		border-radius: if($rounded-corners, $radius-default, 0);
	}

	@if $padding == large {
		@include apply-spacing($size: large);
	}

	@else if $padding == medium {
		@include apply-spacing($size: $padding, $square: true);
	}

	@else if $padding == default {
		@include apply-spacing($size: $padding, $square: true);
	}

	@else {
		padding: 0;
	}

	&:last-child {
		margin-bottom: 0;
	}
}

/// @name card-center-content
@mixin card-center-content() {
	@include display-flex($align: center, $justify: center);

	text-align: center;
}
