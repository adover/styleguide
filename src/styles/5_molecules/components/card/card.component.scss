@import '../../../2_tools/mixins/icon-inline.scss';
@import '../../../2_tools/mixins/position';
@import '../../../4_atoms/card';

:host {
	display: block;

	&[card-center='true'] {
		@include card-center-content();
	}

	&[card-padding='none'] {
		&[card-rounded='true'] {
			@include card-layout($padding: false, $rounded-corners: true);
		}

		&[card-rounded='false'] {
			@include card-layout($padding: false, $rounded-corners: false);
		}

		&[card-rounded='default'] {
			@include card-layout($padding: false, $rounded-corners: default);
		}

		&[card-rounded='large'] {
			@include card-layout($padding: false, $rounded-corners: large);
		}

		&[card-rounded='medium'] {
			@include card-layout($padding: false, $rounded-corners: medium);
		}
	}

	&[card-padding='default'] {
		&[card-rounded='true'] {
			@include card-layout($padding: default, $rounded-corners: true);
		}

		&[card-rounded='false'] {
			@include card-layout($padding: default, $rounded-corners: false);
		}

		&[card-rounded='default'] {
			@include card-layout($padding: default, $rounded-corners: true);
		}

		&[card-rounded='large'] {
			@include card-layout($padding: default, $rounded-corners: large);
		}

		&[card-rounded='medium'] {
			@include card-layout($padding: default, $rounded-corners: medium);
		}
	}

	&[card-padding='medium'] {
		&[card-rounded='true'] {
			@include card-layout($padding: medium, $rounded-corners: true);
		}

		&[card-rounded='false'] {
			@include card-layout($padding: medium, $rounded-corners: false);
		}

		&[card-rounded='default'] {
			@include card-layout($padding: medium, $rounded-corners: true);
		}

		&[card-rounded='large'] {
			@include card-layout($padding: medium, $rounded-corners: large);
		}

		&[card-rounded='medium'] {
			@include card-layout($padding: medium, $rounded-corners: medium);
		}
	}

	&[card-padding='large'] {
		&[card-rounded='true'] {
			@include card-layout($padding: large, $rounded-corners: true);
		}

		&[card-rounded='false'] {
			@include card-layout($padding: large, $rounded-corners: false);
		}

		&[card-rounded='default'] {
			@include card-layout($padding: large, $rounded-corners: true);
		}

		&[card-rounded='large'] {
			@include card-layout($padding: large, $rounded-corners: large);
		}

		&[card-rounded='medium'] {
			@include card-layout($padding: large, $rounded-corners: medium);
		}
	}

	.card {
		&-title {
			@include icon-inline();
		}
	}
}
