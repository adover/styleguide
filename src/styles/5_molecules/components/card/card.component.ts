import { Component, HostBinding, Input, OnInit } from '@angular/core';
import { PaddingSize } from '../../../ui-models';

export class Card {
	title: string;
	url?: string;
	text?: string;
}

// TODO: This needs refactoring to be like the cdx-svg-icon component
@Component({
	selector: 'cdx-card, [cdxCard]',
	template: `
		<ng-content></ng-content>
	`,
	styleUrls: ['./card.component.scss'],
})
export class CardComponent implements OnInit {
	@HostBinding('class.card') mainClass = true;

	@HostBinding('attr.card-padding')
	@Input()
	padding: PaddingSize = PaddingSize.Default;

	@HostBinding('attr.card-center')
	@Input()
	center = false;

	@HostBinding('attr.card-rounded')
	@Input()
	roundedCorners: false | true | 'large' | 'default' = false;

	@Input() url: string;
	@Input() title: string;
	@Input() text: string;

	constructor() {}

	ngOnInit() {}
}
