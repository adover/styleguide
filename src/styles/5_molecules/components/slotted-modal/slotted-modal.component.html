<div class="modal-backdrop overlay--opaque" (click)="close()"></div>
<div class="modal-contentContainer">
    <cdx-card
            cdkTrapFocus
            [cdkTrapFocusAutoCapture]="canDismiss"
            [center]="center"
            [padding]="padding"
            [attr.data-size]="size"
            [attr.data-mobile-full-height]="mobileFullHeight"
            roundedCorners="medium">
        <button cdkFocusInitial aria-label="Dismiss modal" class="modal-closeBtn" *ngIf="canDismiss" (click)="close()">
            <cdx-svg-icon name="cross" size="small" fill="dark"></cdx-svg-icon>
        </button>
        <div class="modal-slotContainer">
            <div class="modal-headerSlot">
                <ng-content select="[modal-header]"></ng-content>
            </div>
            <div class="modal-contentSlot">
                <ng-content select="[modal-content]"></ng-content>
            </div>
            <div class="modal-footerSlot">
                <ng-content select="[modal-footer-primary]"></ng-content>
                <ng-content select="[modal-footer-secondary]"></ng-content>
            </div>
        </div>
    </cdx-card>
</div>