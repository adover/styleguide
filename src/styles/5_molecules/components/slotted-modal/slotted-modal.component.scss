@import '../../../1_settings/z-index';
@import '../../../2_tools/mixins/apply-spacing';
@import '../../../2_tools/mixins/display-flex';
@import '../../../2_tools/mixins/media-query';
@import '../../../2_tools/mixins/position';
@import '../../../5_molecules/modal';

:host {
  .modal {
    &-backdrop {
      @include modal-backdrop();
    }

    // Below is the same modal styling as used in other modals but built mobile first
    &-contentContainer {
      @include display-flex($align: center, $justify: center);

      @include position(fixed) {
        left: 0;
        top: 0;
      }

      height: 100%;
      pointer-events: none;
      width: 100%;
      z-index: $modal-zindex;

      .card {
        @include apply-spacing($axis: vertical, $size: large, $scale: false);

        @include position() {
          bottom: 0;
          left: 0;
        }

        max-height: map-deep-get($config, modal, height, large);
        max-width: map-get($breakpoints, tablet-portrait);
        pointer-events: auto;
        width: 100%;

        @include media-query($size: tablet-portrait) {
          @include position(relative);

          width: auto;

          @include modal-width();
        }

        @include media-query($size: tablet-portrait, $max: true) {
          border-bottom-left-radius: 0;
          border-bottom-right-radius: 0;
        }

        &[data-mobile-full-height='true'] {
          height: map-deep-get($config, modal, height, full);
          max-height: map-deep-get($config, modal, height, full);

          @include media-query($size: tablet-portrait) {
            height: 100%;
            max-height: map-deep-get($config, modal, height, large);
          }
        }
      }
    }

    &-closeBtn {
      @include modal-close-newUi();
    }

    &-loading {
      // Gives the loading container a larger size so the modal size jumps around less when animating in
      min-height: 10rem;

      @include display-flex($justify: center, $align: center)
    }

    &-slotContainer {
      display: flex;
      flex-direction: column;
      height: 100%;
    }

    &-headerSlot, &-footerSlot {
      flex: 0 0 auto;
    }

    &-footerSlot {
      @include display-flex($direction: column);

      @include media-query($size: tablet-portrait) {
        @include display-flex();
      }
    }

    &-contentSlot {
      flex: 1 1 auto;
      overflow-y: auto;

      @include section-margin($size: default);
    }
  }

  // IE11 fix
  &::ng-deep {
    h5,
    p {
      width: 100%;
    }
  }
}