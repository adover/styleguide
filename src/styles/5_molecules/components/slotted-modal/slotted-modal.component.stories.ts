import { storiesOf } from '@storybook/angular';
import { ApiService, ModalOverlayService, StyleguideModule, WINDOW_PROVIDERS } from '@woolworthsnz/styleguide';
import { CurrencyPipe, DecimalPipe } from '@angular/common';
import { ProductModule, ProductService } from '@woolworthsnz/product';
import { Overlay } from '@angular/cdk/overlay';
import { BrowserAnimationsModule } from '@angular/platform-browser/animations';
import { HttpClientTestingModule } from '@angular/common/http/testing';
import { PortalModule } from '@angular/cdk/portal';
import { ReactiveFormsModule } from '@angular/forms';
import { RouterTestingModule } from '@angular/router/testing';
import { centered } from '@storybook/addon-centered/angular';

export const ModalComponentMetadata = {
	declarations: [],
	providers: [ApiService, CurrencyPipe, DecimalPipe, ModalOverlayService, ProductService, Overlay, WINDOW_PROVIDERS],
	imports: [
		BrowserAnimationsModule,
		HttpClientTestingModule,
		PortalModule,
		ReactiveFormsModule,
		RouterTestingModule,
		StyleguideModule,
		ProductModule,
	],
};

storiesOf('Molecules | Slotted Modal', module)
	.addDecorator(centered)
	.add('default', () => {
		return {
			moduleMetadata: ModalComponentMetadata,
			template: `
				<cdx-slotted-modal>
					<h1 modal-header>I'm in the header</h1>
					<div modal-content>
						<p>I'm content</p>
						<p>I'm content</p>
						<p>I'm content</p>
						<p>I'm content</p>
						<p>I'm content</p>
						<p>I'm content</p>
						<p>I'm content</p>
						<p>I'm content</p>
						<p>I'm content</p>
						<p>I'm content</p>
					</div>
					<button cdxButton [fullWidth]="true" fillStyle="primary" modal-footer-primary>I'm primary</button>
					<button cdxButton [fullWidth]="true" fillStyle="secondary" modal-footer-secondary>I'm secondary</button>
				</cdx-slotted-modal>
				<cdx-svg-definitions></cdx-svg-definitions>
			`,
		};
	});
