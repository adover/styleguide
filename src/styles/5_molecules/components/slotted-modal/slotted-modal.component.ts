import { Component, HostListener, Input } from '@angular/core';
import { Router } from '@angular/router';
import { HistoryService } from '../../../services/history.service';
import { ModalSize, PaddingSize } from '../../../ui-models';

@Component({
	selector: 'cdx-slotted-modal',
	templateUrl: './slotted-modal.component.html',
	styleUrls: ['./slotted-modal.component.scss'],
})
export class SlottedModalComponent {
	@Input() size: ModalSize = ModalSize.Full;
	@Input() center = false;
	@Input() padding = PaddingSize.Medium;
	@Input() canDismiss = true;
	@Input() canGoBack = false;
	@Input() mobileFullHeight = false;

	constructor(private historyService: HistoryService, private router: Router) {}

	@HostListener('document:keydown', ['$event'])
	onKeydown(event: KeyboardEvent) {
		// 'Esc' is IE/Edge specific, it's 'Escape' in every other browser
		if (event.key === 'Esc' || event.key === 'Escape') {
			this.close();
			event.preventDefault();
		}
	}

	close() {
		this.router.navigateByUrl(this.historyService.getPreviousNonModalUrl());
	}
}
