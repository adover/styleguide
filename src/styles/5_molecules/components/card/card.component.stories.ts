import { text, withKnobs } from '@storybook/addon-knobs';
import { storiesOf } from '@storybook/angular';
import { CardComponent } from './card.component';

storiesOf('Molecules|Card', module)
	.addDecorator(withKnobs)
	.add('With text', () => {
		const url = text('url', '#someurl');
		const title = text('title', 'Abes Bagels Gluten Free 270g 3pk');
		const content = text('text', 'Lorem Ipsum Dollar  ');

		return {
			moduleMetadata: {
				declarations: [CardComponent],
			},
			template: `
        <cdx-card [url]="url" [text]="content" [title]="title">
          I am a card
        </cdx-card>`,
			props: {
				url,
				title,
				content,
			},
		};
	})
	.add('Without text', () => {
		const url = text('url', '#someurl');
		const title = text('title', 'Abes Bagels Gluten Free 270g 3pk');

		return {
			moduleMetadata: {
				declarations: [CardComponent],
			},
			template: `
        <cdx-card [url]="url"  [title]="title">
          <img src="https://shop.countdown.co.nz/Content/ProductImages/zoom/9421001850783.jpg/Abes-Bagels-Everything-Bagel-360g.jpg"  image />
        </cdx-card>`,
			props: {
				url,
				title,
			},
		};
	});
