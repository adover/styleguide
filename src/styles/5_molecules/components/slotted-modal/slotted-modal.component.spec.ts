import { Component } from '@angular/core';
import { async, ComponentFixture, TestBed } from '@angular/core/testing';
import { RouterTestingModule } from '@angular/router/testing';
import { SvgIconComponent } from '@woolworthsnz/styleguide';
import { MockComponent } from 'ng-mocks';
import { HistoryService } from '../../../services/history.service';
import { CardComponent } from '../card/card.component';
import { SlottedModalComponent } from './slotted-modal.component';
import { A11yModule } from '@angular/cdk/a11y';

@Component({
	template: `
		<cdx-slotted-modal>
			<div modal-content><p>Some content</p></div>
			<span modal-footer-primary>Footer</span>
			<h1 modal-header>Header</h1>
		</cdx-slotted-modal>
	`,
})
class SlottedModalContentComponent {
	constructor() {}
}

describe('SlottedModalComponent', () => {
	let component: SlottedModalComponent;
	let fixture: ComponentFixture<SlottedModalComponent>;
	let componentWithContent: SlottedModalContentComponent;
	let fixtureWithContent: ComponentFixture<SlottedModalContentComponent>;

	beforeEach(async(() => {
		TestBed.configureTestingModule({
			imports: [RouterTestingModule, A11yModule],
			declarations: [
				SlottedModalComponent,
				CardComponent,
				MockComponent(SvgIconComponent),
				SlottedModalContentComponent,
			],
			providers: [HistoryService],
		}).compileComponents();
	}));

	beforeEach(() => {
		fixture = TestBed.createComponent(SlottedModalComponent);
		component = fixture.componentInstance;
		fixture.detectChanges();

		fixtureWithContent = TestBed.createComponent(SlottedModalContentComponent);
		componentWithContent = fixtureWithContent.componentInstance;
		fixtureWithContent.detectChanges();
	});

	it('should create', () => {
		expect(component).toBeTruthy();
	});

	it('should put content in slots', () => {
		expect(fixtureWithContent).toMatchSnapshot();
	});
});
