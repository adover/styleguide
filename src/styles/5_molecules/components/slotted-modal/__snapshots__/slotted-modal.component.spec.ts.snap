// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SlottedModalComponent should put content in slots 1`] = `
<ng-component
  __ngContext__={[Function LRootView]}
>
  <cdx-slotted-modal>
    <div
      class="modal-backdrop overlay--opaque"
    />
    <div
      class="modal-contentContainer"
    >
      <div
        aria-hidden="true"
        class="cdk-visually-hidden cdk-focus-trap-anchor"
        tabindex="0"
      />
      <cdx-card
        card-center="false"
        card-padding="medium"
        card-rounded="medium"
        cdktrapfocus=""
        class="card"
        data-mobile-full-height="false"
        data-size="full"
        roundedcorners="medium"
      >
        <button
          aria-label="Dismiss modal"
          cdkfocusinitial=""
          class="modal-closeBtn"
        >
          <cdx-svg-icon
            fill="dark"
            name="cross"
            size="small"
          />
        </button>
        
        <div
          class="modal-slotContainer"
        >
          <div
            class="modal-headerSlot"
          >
            <h1
              modal-header=""
            >
              Header
            </h1>
          </div>
          <div
            class="modal-contentSlot"
          >
            <div
              modal-content=""
            >
              <p>
                Some content
              </p>
            </div>
          </div>
          <div
            class="modal-footerSlot"
          >
            <span
              modal-footer-primary=""
            >
              Footer
            </span>
          </div>
        </div>
      </cdx-card>
      <div
        aria-hidden="true"
        class="cdk-visually-hidden cdk-focus-trap-anchor"
        tabindex="0"
      />
    </div>
  </cdx-slotted-modal>
</ng-component>
`;
